### PR TITLE
[Core] Nanoflow-style Computation-Communication Overlap

### PR DESCRIPTION
**Pull request by Conless**
# Overview
Following discussions with Woosuk Kwon ([woosuk.kwon@berkeley.edu](mailto:woosuk.kwon@berkeley.edu)), this PR integrates the computation-communication overlap technique from Nanoflow [1] into vLLM using a non-intrusive, compilation-based approach. The integration operates at the **Torch FX graph** level by partitioning input batches into nano-batches and duplicating selected operations to overlap compute and communication operations. Key features include: 
- Non-intrusive code path: This design introduces fewer than 50 lines of changes to the core vLLM codebase and avoids any model-specific modifications. 
- Robust performance gain: It delivers ~10% speedup for certain workloads (e.g., large-batch inference), while guarding against performance regression for small-batch inference via a simple heuristic.

[1] NanoFlow: Towards Optimal Large Language Model Serving Throughput, OSDI 2025

# Design and Implementation

To enable seamless and transparent intra-device parallelism, this PR introduces a graph-level transformation applied during model compilation. The transformation operates on the traced `torch.fx.Graph`, rather than the original model source code, making it entirely transparent to model implementations. This allows for broad applicability across different models and deployment backends.

The figure below illustrates the overall pipeline of our approach:

- **Graph Transformation**: During the compilation phase, the traced `torch.fx.Graph` is passed to a transformation function that partitions the graph into submodules based on resource usage patterns (e.g., computation vs. communication). These submodules are then duplicated to process different parts of the input batch, enabling pipelined execution with overlapping compute and communication. Notably, the graph is input batch size agnostic. The resulting transformed graphs are cached in a split manager to avoid runtime recompilation and thereby minimize CPU overhead.

- **Attention Metadata Preparation**: At run time, the model runner provides input batch information to a context preparation function. This function determines the nano-batch sizes, prepares the necessary attention metadata for each nano-batch, and stores all globally shared data (such as `vllm.ForwardContext` instances for each nano-batch) into the split manager.

- **Run-time Hook**: During execution, the model’s `forward` method is redirected to a custom runtime callable. This callable retrieves the appropriate cached graph module and executes it using custom forward hooks. These hooks dynamically override the global forward context for each nano-batch to ensure correct and efficient execution.

- **Graceful Degradation**: Splitting into nano-batches does incur overheads for small batch inputs. To maintain robustness across varying workloads and avoid GPU underutilization, the system automatically skips nano-batch splitting when the total token batch size is below a threshold (`min_nano_split_tokens`, default: 1024). Additionally, because this is a graph-level optimization, the entire feature can be toggled via a simple configuration flag, ensuring no performance regressions when disabled.

design

# Evaluations

We tested the current implementation with LLaMA 3-8B model on 2xH200 GPUs (TP=2, use_cudagraph=False) by `benchmark_throughput.py`. It reduces the single-iteration latency by 13% and increases the end-to-end throughput by up to 8%.

Input 512 output 0
|                               | Throughput (token/s) |
|--------------------------|-----------------------------|
| vLLM                     | 69533.7                     |
| vLLM w/ Nanoflow | 74532.8                     |

Input 512 output 512
|                               | Throughput (token/s) |
|--------------------------|-----------------------------|
| vLLM                     | 23018.2                     |
| vLLM w/ Nanoflow | 24066.4                     |

Input 1024 output 512
|                               | Throughput (token/s) |
|--------------------------|-----------------------------|
| vLLM                     | 31135.3                     |
| vLLM w/ Nanoflow | 32587.0                     |

# Discussion & Future Work
In the future, we plan to add these features on top of the current design:
- Explore CUDA graph compatibility
- Implement more types of intra-device parallelism, such as overlapping compute- and memory-bound operators and dual-batch overlap for MoE models

Co-authored-by: Kan Zhu <[kanzhu@cs.washington.edu](mailto:kanzhu@cs.washington.edu)>
Co-authored-by: Yilong Zhao <[yilongzhao@berkeley.edu](mailto:yilongzhao@berkeley.edu)>
Co-authored-by: Ziren Wang <[zirenw2@cs.washington.edu](mailto:zirenw2@cs.washington.edu)>
Co-authored-by: Stephanie Wang <[smwang@cs.washington.edu](mailto:smwang@cs.washington.edu)>
Co-authored-by: Baris Kasikci <[baris@cs.washington.edu](mailto:baris@cs.washington.edu)>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/utils/nano_split.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    cumsum_dtype: Optional[np.dtype] = None,
+) -> tuple[np.ndarray, np.ndarray]:
+    cu_num_tokens = np.cumsum(num_tokens, dtype=cumsum_dtype)
+    total_num_tokens = cu_num_tokens[-1]
+    cumsums_offsets = np.repeat(cu_num_tokens - num_tokens, num_tokens)
+    arange = np.arange(total_num_tokens) - cumsums_offsets
+    return cu_num_tokens, arange
+
+
+def prepare_nano_split_and_set_hooks(
```
Comment (LucasWilkinson) [inline vllm/utils/nano_split.py]: We should unify the logic here with: https://github.com/vllm-project/vllm/pull/21153 so there can be shared attention splitting between this and the upcoming MoE dual batch overlap implementation

Comment (Conless) [reply vllm/utils/nano_split.py]: Thanks for the suggestion and this makes sense to me! The main challenge is that the current `split_attn_metadata` interface in that PR takes the original `common_attn_metadata` as input. This forces the splitting logic into `_prepare_inputs`, which couples it tightly with the existing preparation logic and makes the integration more intrusive. There is a few options to unify the logic while keeping things flexible:
1. Add new interfaces that work directly from the scheduler output
2. Have `_prepare_inputs` return the original `common_attn_metadata`
3. Put the original `common_attn_metadata` into the builder-generated metadata, so it can be accessed later through the forward context

Comment (LucasWilkinson) [reply vllm/utils/nano_split.py]: I think move the splitting logic into prepare inputs is fine; base on my understanding for 2+3 we still call `_prepare_inputs` which means duplicating `builder.build` calls, this this is on the hot path (directly impacts TPOT in low-qps regimes) we should be minimizing duplicated work as much as possible. I could potentially see 1 being an option but would likely lead to duplicated code. 

I think micro-batching will become fairly commonly used both through nanoflow and the wide-ep micro batching @SageMoore and I are working on so I think its fine for it to be a first class citizen in the gpu_model_runner. We should have a draft PR up very soon so you can see our planned gpu_model_runner changes :+1: (cc @SageMoore)

File: vllm/compilation/nanoflow/manager.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+from vllm.compilation.nanoflow.split_utils import (FakeModule, NanoOpInfo,
+                                                   NanoSplitConfig,
+                                                   analyze_graph,
+                                                   get_split_config,
+                                                   split_graph, tag_graph)
+from vllm.config import VllmConfig
+
+
+class NanoSplitManager:
```
Comment (zou3519) [inline vllm/compilation/nanoflow/manager.py:24]: >  The integration operates at the Torch FX graph level by partitioning input batches into nano-batches and duplicating selected operations to overlap compute and communication operations.

Could you explain a bit more on what exactly "nanoflow" does? My understanding is that you split the input batch (of tokens) into smaller batches. What exactly is "duplicating selected operations to overlap compute and communication operations." ?

Comment (zou3519) [reply vllm/compilation/nanoflow/manager.py:24]: I found an example over at https://colab.research.google.com/drive/1zpoptkLA0UiW8ZxnBnrVj2MzEpPectQL?authuser=1 , which helps. Some concrete questions:
1) how do you decide what operations to put on which stream?
2) what happens in the case of multiple batches? The example only has two.

File: vllm/utils/nano_split.py (hunk 1 above)
Comment (Conless) [reply vllm/utils/nano_split.py]: Got it. Looking forward to seeing the planned changes!

File: vllm/compilation/nanoflow/manager.py (hunk 2 above)
Comment (Conless) [reply vllm/compilation/nanoflow/manager.py:24]: > how do you decide what operations to put on which stream?

The assignment of operations to streams depends on which resource each operation is bottlenecked by. For example, GEMM operations are assigned to the compute stream, decode attention to the memory stream, and all-reduce to the network stream. This is implemented by matching operator names (e.g., `"vllm.all_reduce"` gets mapped to the network stream).

> what happens in the case of multiple batches? The example only has two.

For cases with more nano-batches, the approach is similar: the input batch is divided into nano-batches, and operators for each nano-batch are launched on streams according to their type (compute, memory, network). Sequential dependencies within each nano-batch are maintained using event synchronization to ensure the correct execution order.

File: vllm/config/__init__.py [hunk 3]
```diff
@@ -501,6 +501,13 @@ class ModelConfig:
     logits_processors: Optional[list[Union[str, type[LogitsProcessor]]]] = None
     """One or more logits processors' fully-qualified class names or class
     definitions"""
+    enable_nano_batch_split: bool = False
+    """Enable splitting the input batch into nano-batches for intra-device
+    parallelism"""
+    max_num_nano_batches: int = 2
+    """Maximum number of nano-batches to split the input batch into"""
+    min_nano_split_tokens: int = 1024
+    """Minimum number of tokens to split the input batch"""
```
Comment (hmellor) [inline vllm/config/__init__.py]: These aren't really model config, should they belong in the compilation config?

Comment (Conless) [reply vllm/config/__init__.py]: Make sense. Let me create a separate section in the compilation config

**Comment by ProExpertProg** [comment]
@Conless is this ready for review again?

**Comment by Conless** [comment]
> @Conless is this ready for review again?

Hi @ProExpertProg , the current version can work well but has some conflict with the latest DBO PRs #23693 #24845 . I'll fix them soon and ping you later!
